Index: linux-2.6.37.1.x/fs/ext4/Kconfig
===================================================================
--- linux-2.6.37.1.x.orig/fs/ext4/Kconfig	2010-12-29 12:40:35.000000000 +0530
+++ linux-2.6.37.1.x/fs/ext4/Kconfig	2010-12-29 12:52:03.000000000 +0530
@@ -290,6 +290,21 @@
 	help
 	  Extra debug prints to trace snapshot usage of buffer credits.
 
+config EXT4_FS_SNAPSHOT_LIST
+	bool "snapshot list support"
+	depends on EXT4_FS_SNAPSHOT_FILE
+	default y
+	help
+	  Implementation of multiple incremental snapshots.
+	  Snapshot inodes are chained on a list starting at the super block,
+	  both on-disk and in-memory, similar to the orphan inodes.  We reuse
+	  the orphan inode list functions to manipulate the snapshot list.
+	  Unlink and truncate of snapshot inodes on the list is not allowed,
+	  so an inode can never be chained on both orphan and snapshot lists.
+	  We make use of this fact to overload the in-memory inode field
+	  ext4_inode_info.i_orphan for the chaining of snapshots.
+
+
 config EXT4_FS_SNAPSHOT_CTL_INIT
 	bool "snapshot control - init new snapshot"
 	depends on EXT4_FS_SNAPSHOT_CTL
Index: linux-2.6.37.1.x/fs/ext4/namei.c
===================================================================
--- linux-2.6.37.1.x.orig/fs/ext4/namei.c	2010-12-28 22:32:22.000000000 +0530
+++ linux-2.6.37.1.x/fs/ext4/namei.c	2010-12-29 12:58:58.000000000 +0530
@@ -1970,9 +1970,30 @@
  * At filesystem recovery time, we walk this list deleting unlinked
  * inodes and truncating linked inodes in ext4_orphan_cleanup().
  */
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+/*
+ * On errors, ext4_std_error() is called and transaction is aborted.
+ */
 int ext4_orphan_add(handle_t *handle, struct inode *inode)
 {
 	struct super_block *sb = inode->i_sb;
+	/*
+	 * only get the field address from the super block structs
+	 * the content of the field will only be changed under lock_super()
+	 */
+	return ext4_inode_list_add(handle, inode, &NEXT_ORPHAN(inode),
+			&EXT4_SB(sb)->s_es->s_last_orphan,
+			&EXT4_SB(sb)->s_orphan, "orphan");
+}
+
+int ext4_inode_list_add(handle_t *handle, struct inode *inode,
+		__u32 *i_next, __le32 *s_last,
+		struct list_head *s_list, const char *name)
+#else
+int ext4_orphan_add(handle_t *handle, struct inode *inode)
+#endif
+{
+	struct super_block *sb = inode->i_sb;
 	struct ext4_iloc iloc;
 	int err = 0, rc;
 
@@ -2015,9 +2036,19 @@
 		(le32_to_cpu(EXT4_SB(sb)->s_es->s_inodes_count)))
 			goto mem_insert;
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	snapshot_debug(4, "add inode %lu to %s list\n",
+			inode->i_ino, name);
+
+	/* Insert this inode at the head of the on-disk inode list... */
+	*i_next = le32_to_cpu(*s_last);
+	*s_last = cpu_to_le32(inode->i_ino);
+#else
+
 	/* Insert this inode at the head of the on-disk orphan list... */
 	NEXT_ORPHAN(inode) = le32_to_cpu(EXT4_SB(sb)->s_es->s_last_orphan);
 	EXT4_SB(sb)->s_es->s_last_orphan = cpu_to_le32(inode->i_ino);
+#endif
 	err = ext4_handle_dirty_metadata(handle, NULL, EXT4_SB(sb)->s_sbh);
 	rc = ext4_mark_iloc_dirty(handle, inode, &iloc);
 	if (!err)
@@ -2033,11 +2064,22 @@
 	 * anyway on the next recovery. */
 mem_insert:
 	if (!err)
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+		list_add(&EXT4_I(inode)->i_orphan, s_list);
+#else
 		list_add(&EXT4_I(inode)->i_orphan, &EXT4_SB(sb)->s_orphan);
+#endif
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	snapshot_debug(4, "last_%s will point to inode %lu\n",
+			name, inode->i_ino);
+	snapshot_debug(4, "%s inode %lu will point to inode %d\n",
+			name, inode->i_ino, *i_next);
+#else
 	jbd_debug(4, "superblock will point to %lu\n", inode->i_ino);
 	jbd_debug(4, "orphan inode %lu will point to %d\n",
 			inode->i_ino, NEXT_ORPHAN(inode));
+#endif
 out_unlock:
 	mutex_unlock(&EXT4_SB(sb)->s_orphan_lock);
 	ext4_std_error(inode->i_sb, err);
@@ -2048,7 +2090,28 @@
  * ext4_orphan_del() removes an unlinked or truncated inode from the list
  * of such inodes stored on disk, because it is finally being cleaned up.
  */
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+int ext4_orphan_del(handle_t *handle, struct inode *inode)
+{
+	struct super_block *sb = inode->i_sb;
+	/*
+	 * only get the field address from the super block structs
+	 * the content of the field will only be changed under lock_super()
+	 */
+	return ext4_inode_list_del(handle, inode, &NEXT_ORPHAN(inode),
+			&EXT4_SB(sb)->s_es->s_last_orphan,
+			&EXT4_SB(sb)->s_orphan, "orphan");
+}
+
+#define NEXT_INODE_OFFSET (((char *)inode)-((char *)i_next))
+#define NEXT_INODE(i_prev) (*(__u32 *)(((char *)i_prev)-NEXT_INODE_OFFSET))
+
+int ext4_inode_list_del(handle_t *handle, struct inode *inode,
+		__u32 *i_next, __le32 *s_last,
+		struct list_head *s_list, const char *name)
+#else
 int ext4_orphan_del(handle_t *handle, struct inode *inode)
+#endif
 {
 	struct list_head *prev;
 	struct ext4_inode_info *ei = EXT4_I(inode);
@@ -2065,11 +2128,20 @@
 	if (list_empty(&ei->i_orphan))
 		goto out;
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	ino_next = *i_next;
+#else
 	ino_next = NEXT_ORPHAN(inode);
+#endif
 	prev = ei->i_orphan.prev;
 	sbi = EXT4_SB(inode->i_sb);
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	snapshot_debug(4, "remove inode %lu from %s list\n", inode->i_ino,
+		       name);
+#else
 	jbd_debug(4, "remove inode %lu from orphan list\n", inode->i_ino);
+#endif
 
 	list_del_init(&ei->i_orphan);
 
@@ -2084,30 +2156,53 @@
 	if (err)
 		goto out_err;
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	if (prev == s_list) {
+		snapshot_debug(4, "last_%s will point to inode %lu\n", name,
+				ino_next);
+#else
 	if (prev == &sbi->s_orphan) {
 		jbd_debug(4, "superblock will point to %u\n", ino_next);
+#endif
 		BUFFER_TRACE(sbi->s_sbh, "get_write_access");
 		err = ext4_journal_get_write_access(handle, sbi->s_sbh);
 		if (err)
 			goto out_brelse;
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+		*s_last = cpu_to_le32(ino_next);
+#else
 		sbi->s_es->s_last_orphan = cpu_to_le32(ino_next);
+#endif
 		err = ext4_handle_dirty_metadata(handle, NULL, sbi->s_sbh);
 	} else {
 		struct ext4_iloc iloc2;
 		struct inode *i_prev =
 			&list_entry(prev, struct ext4_inode_info, i_orphan)->vfs_inode;
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+		snapshot_debug(4, "%s inode %lu will point to inode %lu\n",
+			  name, i_prev->i_ino, ino_next);
+#else
 		jbd_debug(4, "orphan inode %lu will point to %u\n",
 			  i_prev->i_ino, ino_next);
+#endif
 		err = ext4_reserve_inode_write(handle, i_prev, &iloc2);
 		if (err)
 			goto out_brelse;
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+		NEXT_INODE(i_prev) = ino_next;
+#else
 		NEXT_ORPHAN(i_prev) = ino_next;
+#endif
 		err = ext4_mark_iloc_dirty(handle, i_prev, &iloc2);
 	}
 	if (err)
 		goto out_brelse;
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	*i_next = 0;
+#else
 	NEXT_ORPHAN(inode) = 0;
+#endif
 	err = ext4_mark_iloc_dirty(handle, inode, &iloc);
 
 out_err:
Index: linux-2.6.37.1.x/fs/ext4/snapshot_ctl.c
===================================================================
--- linux-2.6.37.1.x.orig/fs/ext4/snapshot_ctl.c	2010-12-29 12:40:35.000000000 +0530
+++ linux-2.6.37.1.x/fs/ext4/snapshot_ctl.c	2010-12-29 13:01:59.000000000 +0530
@@ -390,12 +390,28 @@
 #endif
 #endif
 	ext4_fsblk_t snapshot_blocks = ext4_blocks_count(sbi->s_es);
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	struct list_head *l, *list = &sbi->s_snapshot_list;
+
+	if (!list_empty(list)) {
+		struct inode *last_snapshot =
+			&list_first_entry(list, struct ext4_inode_info,
+					  i_snaplist)->vfs_inode;
+		if (active_snapshot != last_snapshot) {
+			snapshot_debug(1, "failed to add snapshot because last"
+				       " snapshot (%u) is not active\n",
+				       last_snapshot->i_generation);
+			return -EINVAL;
+		}
+	}
+#else
 	if (active_snapshot) {
 		snapshot_debug(1, "failed to add snapshot because active "
 			       "snapshot (%u) has to be deleted first\n",
 			       active_snapshot->i_generation);
 		return -EINVAL;
 	}
+#endif
 
 	/* prevent take of unlinked snapshot file */
 	if (!inode->i_nlink) {
@@ -458,6 +474,31 @@
 		EXT4_SET_RO_COMPAT_FEATURE(sb,
 			EXT4_FEATURE_RO_COMPAT_HAS_SNAPSHOT);
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	/* add snapshot list reference */
+	if (!igrab(inode)) {
+		err = -EIO;
+		goto out_handle;
+	}
+	/*
+	 * First, the snapshot is added to the in-memory and on-disk list.
+	 * At the end of snapshot_take(), it will become the active snapshot
+	 * in-memory and on-disk.
+	 * Finally, if snapshot_create() or snapshot_take() has failed,
+	 * snapshot_update() will remove it from the in-memory and on-disk list.
+	 */
+	err = ext4_inode_list_add(handle, inode, &NEXT_SNAPSHOT(inode),
+			&sbi->s_es->s_snapshot_list,
+			list, "snapshot");
+	/* add snapshot list reference */
+	if (err) {
+		snapshot_debug(1, "failed to add snapshot (%u) to list\n",
+			       inode->i_generation);
+		iput(inode);
+		goto out_handle;
+	}
+	l = list->next;
+#else
 	lock_super(sb);
 	err = ext4_journal_get_write_access(handle, sbi->s_sbh);
 	sbi->s_es->s_snapshot_list = cpu_to_le32(inode->i_ino);
@@ -466,6 +507,7 @@
 	unlock_super(sb);
 	if (err)
 		goto out_handle;
+#endif
 
 	err = ext4_mark_inode_dirty(handle, inode);
 	if (err)
@@ -620,12 +662,21 @@
 		goto out_handle;
 	}
 #ifdef CONFIG_EXT4_FS_SNAPSHOT_CTL_FIX
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	if (l != list) {
+		ino = list_entry(l, struct ext4_inode_info,
+				i_snaplist)->vfs_inode.i_ino;
+		l = l->next;
+		goto alloc_inode_blocks;
+	}
+#else
 	if (ino == EXT4_ROOT_INO) {
 		ino = inode->i_ino;
 		goto alloc_inode_blocks;
 	}
 #endif
 #endif
+#endif
 
 	snapshot_debug(1, "snapshot (%u) created\n", inode->i_generation);
 	err = 0;
@@ -719,6 +770,10 @@
  */
 int ext4_snapshot_take(struct inode *inode)
 {
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	struct list_head *list = &EXT4_SB(inode->i_sb)->s_snapshot_list;
+	struct list_head *l = list->next;
+#endif
 	struct super_block *sb = inode->i_sb;
 	struct ext4_sb_info *sbi = EXT4_SB(sb);
 	struct ext4_super_block *es = NULL;
@@ -914,12 +969,21 @@
 	mark_buffer_dirty(sbh);
 	sync_dirty_buffer(sbh);
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	if (l != list) {
+		curr_inode = &list_entry(l, struct ext4_inode_info,
+				       i_snaplist)->vfs_inode;
+		l = l->next;
+		goto copy_inode_blocks;
+	}
+#else
 	if (curr_inode->i_ino == EXT4_ROOT_INO) {
 		curr_inode = inode;
 		goto copy_inode_blocks;
 	}
 #endif
 #endif
+#endif
 
 	/* reset COW bitmap cache */
 	err = ext4_snapshot_reset_bitmap_cache(sb, 0);
@@ -1112,6 +1176,16 @@
 	if (err)
 		goto out_handle;
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	err = ext4_inode_list_del(handle, inode, &NEXT_SNAPSHOT(inode),
+			&sbi->s_es->s_snapshot_list,
+			&EXT4_SB(inode->i_sb)->s_snapshot_list,
+			"snapshot");
+	if (err)
+		goto out_handle;
+	/* remove snapshot list reference - taken on snapshot_create() */
+	iput(inode);
+#else
 	lock_super(inode->i_sb);
 	err = ext4_journal_get_write_access(handle, sbi->s_sbh);
 	sbi->s_es->s_snapshot_list = 0;
@@ -1120,6 +1194,7 @@
 	unlock_super(inode->i_sb);
 	if (err)
 		goto out_handle;
+#endif
 	/*
 	 * At this point, this snapshot is empty and not on the snapshots list.
 	 * As long as it was on the list it had to have the LIST flag to prevent
@@ -1185,6 +1260,12 @@
 	int err, num = 0, snapshot_id = 0;
 	int has_active = 0;
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	if (!list_empty(&EXT4_SB(sb)->s_snapshot_list)) {
+		snapshot_debug(1, "warning: snapshots already loaded!\n");
+		return -EINVAL;
+	}
+#endif
 
 #ifdef CONFIG_EXT4_FS_SNAPSHOT_FILE_OLD
 	/* Migrate super block on-disk format */
@@ -1309,8 +1390,15 @@
 			has_active = 1;
 		}
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+		list_add_tail(&EXT4_I(inode)->i_snaplist,
+			      &EXT4_SB(sb)->s_snapshot_list);
+		load_ino = NEXT_SNAPSHOT(inode);
+		/* keep snapshot list reference */
+#else
 		iput(inode);
 		break;
+#endif
 	}
 
 	if (err) {
@@ -1337,6 +1425,18 @@
  */
 void ext4_snapshot_destroy(struct super_block *sb)
 {
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	struct list_head *l, *n;
+	/* iterate safe because we are deleting from list and freeing the
+	 * inodes */
+	list_for_each_safe(l, n, &EXT4_SB(sb)->s_snapshot_list) {
+		struct inode *inode = &list_entry(l, struct ext4_inode_info,
+						  i_snaplist)->vfs_inode;
+		list_del_init(&EXT4_I(inode)->i_snaplist);
+		/* remove snapshot list reference */
+		iput(inode);
+	}
+#endif
 	/* deactivate in-memory active snapshot - cannot fail */
 	(void) ext4_snapshot_set_active(sb, NULL);
 }
@@ -1358,6 +1458,13 @@
 	struct inode *used_by = NULL; /* last non-deleted snapshot found */
 	int deleted;
 #endif
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	struct inode *inode;
+	struct ext4_inode_info *ei;
+	int found_active = 0;
+	int found_enabled = 0;
+	struct list_head *prev;
+#endif
 	int err = 0;
 
 	BUG_ON(read_only && cleanup);
@@ -1365,6 +1472,75 @@
 		EXT4_I(active_snapshot)->i_flags |=
 			EXT4_SNAPFILE_ACTIVE_FL|EXT4_SNAPFILE_LIST_FL;
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	/* iterate safe from oldest snapshot backwards */
+	prev = EXT4_SB(sb)->s_snapshot_list.prev;
+	if (list_empty(prev))
+		return 0;
+
+update_snapshot:
+	ei = list_entry(prev, struct ext4_inode_info, i_snaplist);
+	inode = &ei->vfs_inode;
+	prev = ei->i_snaplist.prev;
+
+	/* all snapshots on the list have the LIST flag */
+	ei->i_flags |= EXT4_SNAPFILE_LIST_FL;
+	/* set the 'No_Dump' flag on all snapshots */
+	ei->i_flags |= EXT4_NODUMP_FL;
+
+	/*
+	 * snapshots later than active (failed take) should be removed.
+	 * no active snapshot means failed first snapshot take.
+	 */
+	if (found_active || !active_snapshot) {
+		if (!read_only)
+			err = ext4_snapshot_remove(inode);
+		goto prev_snapshot;
+	}
+
+	deleted = ei->i_flags & EXT4_SNAPFILE_DELETED_FL;
+	if (!deleted && read_only)
+		/* auto enable snapshots on readonly mount */
+		ext4_snapshot_enable(inode);
+
+	/*
+	 * after completion of a snapshot management operation,
+	 * only the active snapshot can have the ACTIVE flag
+	 */
+	if (inode == active_snapshot) {
+		ei->i_flags |= EXT4_SNAPFILE_ACTIVE_FL;
+		found_active = 1;
+		deleted = 0;
+	} else
+		ei->i_flags &= ~EXT4_SNAPFILE_ACTIVE_FL;
+
+	if (found_enabled)
+		/* snapshot is in use by an older enabled snapshot */
+		ei->i_flags |= EXT4_SNAPFILE_INUSE_FL;
+	else
+		/* snapshot is not in use by older enabled snapshots */
+		ei->i_flags &= ~EXT4_SNAPFILE_INUSE_FL;
+
+	if (cleanup && deleted && !used_by)
+		/* remove permanently unused deleted snapshot */
+		err = ext4_snapshot_remove(inode);
+
+	if (!deleted) {
+		if (!found_active)
+			/* newer snapshots are potentially used by
+			 * this snapshot (when it is enabled) */
+			used_by = inode;
+		if (ei->i_flags & EXT4_SNAPFILE_ENABLED_FL)
+			found_enabled = 1;
+	}
+
+prev_snapshot:
+	if (err)
+		return err;
+	/* update prev snapshot */
+	if (prev != &EXT4_SB(sb)->s_snapshot_list)
+		goto update_snapshot;
+#endif
 
 #ifdef CONFIG_EXT4_FS_SNAPSHOT_CTL
 	if (!active_snapshot || !cleanup || used_by)
Index: linux-2.6.37.1.x/fs/ext4/snapshot.h
===================================================================
--- linux-2.6.37.1.x.orig/fs/ext4/snapshot.h	2010-12-29 12:35:55.000000000 +0530
+++ linux-2.6.37.1.x/fs/ext4/snapshot.h	2010-12-29 12:51:00.000000000 +0530
@@ -347,6 +347,14 @@
 					     unsigned int block_group);
 
 /* namei.c */
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+extern int ext4_inode_list_add(handle_t *handle, struct inode *inode,
+				__u32 *i_next, __le32 *s_last,
+				struct list_head *s_list, const char *name);
+extern int ext4_inode_list_del(handle_t *handle, struct inode *inode,
+				__u32 *i_next, __le32 *s_last,
+				struct list_head *s_list, const char *name);
+#endif
 
 /* inode.c */
 extern ext4_fsblk_t ext4_get_inode_block(struct super_block *sb,
Index: linux-2.6.37.1.x/fs/ext4/super.c
===================================================================
--- linux-2.6.37.1.x.orig/fs/ext4/super.c	2010-12-29 12:39:47.000000000 +0530
+++ linux-2.6.37.1.x/fs/ext4/super.c	2010-12-29 12:51:00.000000000 +0530
@@ -3650,6 +3650,9 @@
 #ifdef CONFIG_EXT4_FS_SNAPSHOT_FILE
 	mutex_init(&sbi->s_snapshot_mutex);
 	sbi->s_active_snapshot = NULL;
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	INIT_LIST_HEAD(&sbi->s_snapshot_list); /* snapshot files */
+#endif
 
 #endif
 	needs_recovery = (es->s_last_orphan != 0 ||
Index: linux-2.6.37.1.x/fs/ext4/ext4.h
===================================================================
--- linux-2.6.37.1.x.orig/fs/ext4/ext4.h	2010-12-29 12:52:46.000000000 +0530
+++ linux-2.6.37.1.x/fs/ext4/ext4.h	2010-12-29 13:00:23.000000000 +0530
@@ -1260,6 +1260,9 @@
 	struct ext4_group_info *s_snapshot_group_info;	/* [ sb_bgl_lock ] */
 	struct mutex s_snapshot_mutex;		/* protects 2 fields below: */
 	struct inode *s_active_snapshot;	/* [ s_snapshot_mutex ] */
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_LIST
+	struct list_head s_snapshot_list;	/* [ s_snapshot_mutex ] */
+#endif
 #endif
 #ifdef CONFIG_JBD2_DEBUG
 	struct timer_list turn_ro_timer;	/* For turning read-only (crash simulation) */
