Index: backup/fs/ext4/inode.c
===================================================================
--- backup.orig/fs/ext4/inode.c	2010-12-02 09:51:38.000000000 +0530
+++ backup/fs/ext4/inode.c	2010-12-02 11:07:15.000000000 +0530
@@ -857,7 +857,12 @@
 	 */
 	if (where->bh) {
 		BUFFER_TRACE(where->bh, "get_write_access");
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+		err = ext4_journal_get_write_access_inode(handle, inode,
+							   where->bh);
+#else
 		err = ext4_journal_get_write_access(handle, where->bh);
+#endif
 		if (err)
 			goto err_out;
 	}
@@ -4395,7 +4400,11 @@
 					    blocks_for_truncate(inode));
 		if (bh) {
 			BUFFER_TRACE(bh, "retaking write access");
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+			ext4_journal_get_write_access_inode(handle, inode, bh);
+#else
 			ext4_journal_get_write_access(handle, bh);
+#endif
 		}
 	}
 
@@ -4441,7 +4450,12 @@
 
 	if (this_bh) {				/* For indirect block */
 		BUFFER_TRACE(this_bh, "get_write_access");
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+		err = ext4_journal_get_write_access_inode(handle, inode,
+							   this_bh);
+#else
 		err = ext4_journal_get_write_access(handle, this_bh);
+#endif
 		/* Important: if we can't update the indirect pointers
 		 * to the blocks, we can't free them. */
 		if (err)
@@ -4599,8 +4613,13 @@
 				 * pointed to by an indirect block: journal it
 				 */
 				BUFFER_TRACE(parent_bh, "get_write_access");
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+				if (!ext4_journal_get_write_access_inode(
+					    handle, inode, parent_bh)){
+#else
 				if (!ext4_journal_get_write_access(handle,
 								   parent_bh)){
+#endif
 					*p = 0;
 					BUFFER_TRACE(parent_bh,
 					"call ext4_handle_dirty_metadata");
Index: backup/fs/ext4/Kconfig
===================================================================
--- backup.orig/fs/ext4/Kconfig	2010-12-02 11:05:44.000000000 +0530
+++ backup/fs/ext4/Kconfig	2010-12-02 11:07:15.000000000 +0530
@@ -98,3 +98,13 @@
 
 	  If you select Y here, then you will be able to turn on debugging
 	  with a command such as "echo 1 > /sys/kernel/debug/ext4/mballoc-debug"
+
+config EXT4_FS_SNAPSHOT_HOOKS_JBD
+	bool "snapshot hooks - inside JBD hooks"
+	depends on EXT4_FS_SNAPSHOT
+	default y
+	help
+	  Before every meta data buffer write, the journal API is called,
+	  namely, one of the ext4_journal_get_XXX_access() functions.
+	  We use these journal hooks to call the snapshot API,
+	  before every meta data buffer write.
Index: backup/fs/ext4/snapshot.h
===================================================================
--- backup.orig/fs/ext4/snapshot.h	2010-12-02 11:05:44.000000000 +0530
+++ backup/fs/ext4/snapshot.h	2010-12-02 11:07:15.000000000 +0530
@@ -107,6 +107,57 @@
  * Block access functions
  */
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+/*
+ * get_write_access() is called before writing to a metadata block
+ * if @inode is not NULL, then this is an inode's indirect block
+ * otherwise, this is a file system global metadata block
+ *
+ * Return values:
+ * = 0 - block was COWed or doesn't need to be COWed
+ * < 0 - error
+ */
+static inline int ext4_snapshot_get_write_access(handle_t *handle,
+		struct inode *inode, struct buffer_head *bh)
+{
+	return ext4_snapshot_cow(handle, inode, bh, 1);
+}
+
+/*
+ * called from ext4_journal_get_undo_access(),
+ * which is called for group bitmap block from:
+ * 1. ext4_free_blocks_sb_inode() before deleting blocks
+ * 2. ext4_new_blocks() before allocating blocks
+ *
+ * Return values:
+ * = 0 - block was COWed or doesn't need to be COWed
+ * < 0 - error
+ */
+static inline int ext4_snapshot_get_undo_access(handle_t *handle,
+		struct buffer_head *bh)
+{
+	return ext4_snapshot_cow(handle, NULL, bh, 1);
+}
+
+/*
+ * get_create_access() is called after allocating a new metadata block
+ *
+ * Return values:
+ * = 0 - block was COWed or doesn't need to be COWed
+ * < 0 - error
+ */
+static inline int ext4_snapshot_get_create_access(handle_t *handle,
+		struct buffer_head *bh)
+{
+	/*
+	 * This block shouldn't need to be COWed if get_delete_access() was
+	 * called for all deleted blocks.  However, it may need to be COWed
+	 * if fsck was run and if it had freed some blocks without moving them
+	 * to snapshot.  In the latter case, -EIO will be returned.
+	 */
+	return ext4_snapshot_cow(handle, NULL, bh, 0);
+}
+#endif
 
 /*
  * Snapshot constructor/destructor
Index: backup/fs/ext4/ext4_jbd2.c
===================================================================
--- backup.orig/fs/ext4/ext4_jbd2.c	2010-12-02 09:51:38.000000000 +0530
+++ backup/fs/ext4/ext4_jbd2.c	2010-12-02 11:16:07.000000000 +0530
@@ -17,6 +17,10 @@
 
 	if (ext4_handle_valid(handle)) {
 		err = jbd2_journal_get_undo_access(handle, bh);
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+	if (!err)
+		err = ext4_snapshot_get_undo_access(handle, bh);
+#endif
 		if (err)
 			ext4_journal_abort_handle(where, line, __func__, bh,
 						  handle, err);
@@ -24,13 +28,23 @@
 	return err;
 }
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+int __ext4_journal_get_write_access_inode(const char *where, unsigned int line,
+					 handle_t *handle, struct inode *inode,
+					 struct buffer_head *bh)
+#else
 int __ext4_journal_get_write_access(const char *where, unsigned int line,
 				    handle_t *handle, struct buffer_head *bh)
+#endif
 {
 	int err = 0;
 
 	if (ext4_handle_valid(handle)) {
 		err = jbd2_journal_get_write_access(handle, bh);
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+	if (!err)
+		err = ext4_snapshot_get_write_access(handle, inode, bh);
+#endif
 		if (err)
 			ext4_journal_abort_handle(where, line, __func__, bh,
 						  handle, err);
@@ -112,6 +127,10 @@
 
 	if (ext4_handle_valid(handle)) {
 		err = jbd2_journal_get_create_access(handle, bh);
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+	if (!err)
+		err = ext4_snapshot_get_create_access(handle, bh);
+#endif
 		if (err)
 			ext4_journal_abort_handle(where, line, __func__,
 						  bh, handle, err);
Index: backup/fs/ext4/ext4_jbd2.h
===================================================================
--- backup.orig/fs/ext4/ext4_jbd2.h	2010-12-02 11:05:44.000000000 +0530
+++ backup/fs/ext4/ext4_jbd2.h	2010-12-02 11:15:36.000000000 +0530
@@ -133,9 +134,16 @@
 int __ext4_journal_get_undo_access(const char *where, unsigned int line,
 				   handle_t *handle, struct buffer_head *bh);
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+int __ext4_journal_get_write_access_inode(const char *where, unsigned int line,
+					 handle_t *handle, struct inode *inode,
+					 struct buffer_head *bh);
+#else
 int __ext4_journal_get_write_access(const char *where, unsigned int line,
 				    handle_t *handle, struct buffer_head *bh);
 
+#endif
+
 int __ext4_forget(const char *where, unsigned int line, handle_t *handle,
 		  int is_metadata, struct inode *inode,
 		  struct buffer_head *bh, ext4_fsblk_t blocknr);
@@ -152,8 +161,18 @@
 
 #define ext4_journal_get_undo_access(handle, bh) \
 	__ext4_journal_get_undo_access(__func__, __LINE__, (handle), (bh))
+
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_HOOKS_JBD
+#define ext4_journal_get_write_access(handle, bh) \
+	__ext4_journal_get_write_access_inode(__func__, __LINE__, \
+						 (handle), NULL, (bh))
+#define ext4_journal_get_write_access_inode(handle, inode, bh) \
+	__ext4_journal_get_write_access_inode(__func__, __LINE__, \
+						(handle), (inode), (bh))
+#else
 #define ext4_journal_get_write_access(handle, bh) \
 	__ext4_journal_get_write_access(__func__, __LINE__, (handle), (bh))
+#endif
 #define ext4_forget(handle, is_metadata, inode, bh, block_nr) \
 	__ext4_forget(__func__, __LINE__, (handle), (is_metadata), (inode), \
 		      (bh), (block_nr))
