diff -Nuarp a/fs/ext4/Kconfig b/fs/ext4/Kconfig
--- a/fs/ext4/Kconfig	2010-11-11 10:47:20.000000000 +0200
+++ b/fs/ext4/Kconfig	2010-11-11 10:47:19.000000000 +0200
@@ -342,3 +342,16 @@ config EXT4_FS_SNAPSHOT_JOURNAL_BYPASS
 	  would further increase the transaction's COW credits factor.
 	  Instead, we choose to pay a small performance penalty on these few
 	  COW bitmap operations and wait until they are synced to disk.
+
+config EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+	bool "snapshot journaled - cache last COW tid in journal_head"
+	depends on EXT4_FS_SNAPSHOT_JOURNAL
+	default y
+	help
+	  Cache last COW transaction id in buffer's journal_head.
+	  The cache suppresses COW tests until the transaction in committed.
+	  By default, the running transaction is committed every 5 seconds,
+	  which implies an average COW cache expiry of 2.5 seconds.
+	  Before taking a new snapshot, the journal is flushed to disk
+	  and the current transaction in committed, so the COW cache is
+	  invalidated (as it should be).
diff -Nuarp a/fs/ext4/snapshot.c b/fs/ext4/snapshot.c
--- a/fs/ext4/snapshot.c	2010-11-11 10:47:21.000000000 +0200
+++ b/fs/ext4/snapshot.c	2010-11-11 10:47:20.000000000 +0200
@@ -502,6 +502,120 @@ __ext4_snapshot_trace_cow(const char *w
 #define ext4_snapshot_trace_cow(where, handle, sb, inode, bh, block, cmd)
 #endif
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+/*
+ * The last transaction ID during which the buffer has been COWed is stored in
+ * the b_cow_tid field of the journal_head struct.  If we know that the buffer
+ * was COWed during the current transaction, we don't need to COW it again.
+ * Find the offset of the b_cow_tid field by checking for free space in the
+ * journal_head struct. If there is no free space, don't use cow tid cache.
+ * This method is used so the ext4 module could be loaded without rebuilding
+ * the kernel with modified journal_head struct.
+ * [jbd_lock_bh_state()]
+ */
+
+static int cow_tid_offset;
+
+void init_ext4_snapshot_cow_cache(void)
+{
+	const struct journal_head *jh = NULL;
+	char *pos, *end;
+
+	if (cow_tid_offset)
+		return;
+
+#ifdef CONFIG_64BIT
+	/* check for 32bit padding to 64bit alignment after b_modified */
+	pos = (char *)&jh->b_modified + sizeof(jh->b_modified);
+	end = (char *)&jh->b_frozen_data;
+	if (pos + sizeof(tid_t) <= end)
+		goto found;
+
+#endif
+	/* check for extra jbd2 fields after last jbd field */
+	pos = (char *)&jh->b_cpprev + sizeof(jh->b_cpprev);
+	end = (char *)jh + sizeof(*jh);
+	if (pos + sizeof(tid_t) <= end)
+		goto found;
+
+	/* no free space found - disable cow cache */
+	cow_tid_offset = -1;
+	return;
+found:
+	cow_tid_offset = pos - (char *)NULL;
+#ifdef CONFIG_EXT4_FS_DEBUG
+	cow_cache_offset = cow_tid_offset;
+#endif
+}
+
+#define cow_cache_enabled()	(cow_tid_offset > 0)
+#define jh_cow_tid(jh)		\
+	*(tid_t *)(((char *)(jh))+cow_tid_offset)
+
+#define test_cow_tid(jh, handle)	\
+	(jh_cow_tid(jh) == (handle)->h_transaction->t_tid)
+#define set_cow_tid(jh, handle)		\
+	jh_cow_tid(jh) = (handle)->h_transaction->t_tid
+
+/*
+ * Journal COW cache functions.
+ * a block can only be COWed once per snapshot,
+ * so a block can only be COWed once per transaction,
+ * so a buffer that was COWed in the current transaction,
+ * doesn't need to be COWed.
+ *
+ * Return values:
+ * 1 - block was COWed in current transaction
+ * 0 - block wasn't COWed in current transaction
+ */
+static int
+ext4_snapshot_test_cowed(handle_t *handle, struct buffer_head *bh)
+{
+	struct journal_head *jh;
+
+	if (!cow_cache_enabled())
+		return 0;
+
+	/* check the COW tid in the journal head */
+	if (bh && buffer_jbd(bh)) {
+		jbd_lock_bh_state(bh);
+		jh = bh2jh(bh);
+		if (jh && !test_cow_tid(jh, handle))
+			jh = NULL;
+		jbd_unlock_bh_state(bh);
+		if (jh)
+			/*
+			 * Block was already COWed in the running transaction,
+			 * so we don't need to COW it again.
+			 */
+			return 1;
+	}
+	return 0;
+}
+
+static void
+ext4_snapshot_mark_cowed(handle_t *handle, struct buffer_head *bh)
+{
+	struct journal_head *jh;
+
+	if (!cow_cache_enabled())
+		return;
+
+	if (bh && buffer_jbd(bh)) {
+		jbd_lock_bh_state(bh);
+		jh = bh2jh(bh);
+		if (jh && !test_cow_tid(jh, handle))
+			/*
+			 * this is the first time this block was COWed
+			 * in the running transaction.
+			 * update the COW tid in the journal head
+			 * to mark that this block doesn't need to be COWed.
+			 */
+			set_cow_tid(jh, handle);
+		jbd_unlock_bh_state(bh);
+	}
+}
+#endif
 
 /*
  * Begin COW or move operation.
@@ -584,6 +698,15 @@ int ext4_snapshot_test_and_cow(const ch
 		return -EPERM;
 	}
 
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+	/* check if the buffer was COWed in the current transaction */
+	if (ext4_snapshot_test_cowed(handle, bh)) {
+		snapshot_debug_hl(4, "buffer found in COW cache - "
+				  "skip block cow!\n");
+		trace_cow_inc(handle, ok_jh);
+		return 0;
+	}
+#endif
 
 	/* BEGIN COWing */
 	ext4_snapshot_cow_begin(handle);
@@ -680,6 +803,10 @@ int ext4_snapshot_test_and_cow(const ch
 test_pending_cow:
 
 cowed:
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+	/* mark the buffer COWed in the current transaction */
+	ext4_snapshot_mark_cowed(handle, bh);
+#endif
 out:
 	brelse(sbh);
 	/* END COWing */
diff -Nuarp a/fs/ext4/snapshot_debug.c b/fs/ext4/snapshot_debug.c
--- a/fs/ext4/snapshot_debug.c	2010-11-11 10:47:21.000000000 +0200
+++ b/fs/ext4/snapshot_debug.c	2010-11-11 10:47:20.000000000 +0200
@@ -46,11 +46,17 @@ static const char *snapshot_test_names[S
 
 u16 snapshot_enable_test[SNAPSHOT_TESTS_NUM] __read_mostly = {0};
 u8 snapshot_enable_debug __read_mostly = 1;
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+u8 cow_cache_offset __read_mostly = 0;
+#endif
 
 static struct dentry *ext4_debugfs_dir;
 static struct dentry *snapshot_debug;
 static struct dentry *snapshot_version;
 static struct dentry *snapshot_test[SNAPSHOT_TESTS_NUM];
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+static struct dentry *cow_cache;
+#endif
 
 static char snapshot_version_str[] = EXT4_SNAPSHOT_VERSION;
 static struct debugfs_blob_wrapper snapshot_version_blob = {
@@ -80,6 +86,11 @@ void ext4_create_debugfs_entry(void)
 					      S_IRUGO|S_IWUSR,
 					      ext4_debugfs_dir,
 					      &snapshot_enable_test[i]);
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+	cow_cache = debugfs_create_u8("cow-cache", S_IRUGO,
+					   ext4_debugfs_dir,
+					   &cow_cache_offset);
+#endif
 }
 
 /*
@@ -92,6 +103,10 @@ void ext4_remove_debugfs_entry(void)
 
 	if (!ext4_debugfs_dir)
 		return;
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+	if (cow_cache)
+		debugfs_remove(cow_cache);
+#endif
 	for (i = 0; i < SNAPSHOT_TESTS_NUM && i < SNAPSHOT_TEST_NAMES; i++)
 		if (snapshot_test[i])
 			debugfs_remove(snapshot_test[i]);
diff -Nuarp a/fs/ext4/snapshot.h b/fs/ext4/snapshot.h
--- a/fs/ext4/snapshot.h	2010-11-11 10:47:21.000000000 +0200
+++ b/fs/ext4/snapshot.h	2010-11-11 10:47:20.000000000 +0200
@@ -314,6 +314,9 @@ extern int ext4_snapshot_get_inode_acce
 					   int count, int cmd,
 					   struct inode **prev_snapshot);
 #endif
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+extern void init_ext4_snapshot_cow_cache(void);
+#endif
 
 /*
  * Snapshot constructor/destructor
@@ -327,6 +330,9 @@ extern void ext4_snapshot_destroy(struc
 static inline int init_ext4_snapshot(void)
 {
 	init_ext4_snapshot_debug();
+#ifdef CONFIG_EXT4_FS_SNAPSHOT_JOURNAL_CACHE
+	init_ext4_snapshot_cow_cache();
+#endif
 	return 0;
 }
 
